File: leetcode_962/MaximumWidthRamp.java
package leetcode_962;

import java.util.Stack;

class MaximumWidthRamp {
    public int maxWidthRamp(int[] nums) {
        int n = nums.length;
        Stack<Integer> stack = new Stack<>();

        // Build a decreasing stack of indices
        for (int i = 0; i < n; i++) {
            if (stack.isEmpty() || nums[stack.peek()] > nums[i]) {
                stack.push(i);
            }
        }

        int maxWidth = 0;

        // Iterate through the array in reverse
        for (int j = n - 1; j >= 0; j--) {
            while (!stack.isEmpty() && nums[stack.peek()] <= nums[j]) {
                maxWidth = Math.max(maxWidth, j - stack.pop());
            }
        }

        return maxWidth;
    }

    public static void main(String[] args) {
        MaximumWidthRamp m = new MaximumWidthRamp();
        System.out.println(m.maxWidthRamp(new int[] { 6, 0, 8, 2, 1, 5 }));
    }
}

File: leetcode_962/MaximumWidthRamp.py
class MaximumWidthRamp:
    def max_width_ramp(self, nums: list[int]) -> int:
        stack = []
        res = 0
        # Build a decreasing stack of indices
        for i, a in enumerate(nums):
            if not stack or nums[stack[-1]] > a:
                stack.append(i)

        # Iterate through the array in reverse
        for i in range(len(nums) - 1, -1, -1):
            while stack and nums[stack[-1]] <= nums[i]:
                res = max(res, i - stack.pop())
        return res


def main():
    main_instance = MaximumWidthRamp()
    print(main_instance.max_width_ramp([6, 0, 8, 2, 1, 5]))


if __name__ == "__main__":
    main()
